keystore/keystore.go

```diff
@@ -0,0 +1,222 @@
+		N: gethkeystore.StandardScryptN,
+		P: gethkeystore.StandardScryptP,
+	}
+	FastScryptParams ScryptParams = ScryptParams{
+		N: 1 << 14,
+		P: 1,
+	}
+)
+
+// KeyInfo is the information about a key in the keystore.
+// Public key may be empty for non-asymmetric key types.
+type KeyInfo struct {
+	Name      string
+	KeyType   KeyType
+	CreatedAt time.Time
+	PublicKey []byte
+	Metadata  []byte
+}
+
+// newKeyInfo is a private constructor that ensures all fields are set explicitly.
+func newKeyInfo(name string, keyType KeyType, createdAt time.Time, publicKey []byte, metadata []byte) KeyInfo {
+	return KeyInfo{
+		Name:      name,
+		KeyType:   keyType,
+		CreatedAt: createdAt,
+		PublicKey: publicKey,
+		Metadata:  metadata,
+	}
+}
+
+type Keystore interface {
+	Admin
+	Reader
+	Signer
+	Encryptor
+}
+
+type key struct {
+	keyType    KeyType
+	privateKey internal.Raw
+	createdAt  time.Time
+	metadata   []byte
+
+	// Derived from private key on loading from storage.
+	// Cached here for convenience.
+	publicKey []byte
+}
+
+// newKey is a private constructor ensuring the internal key struct is fully and correctly populated.
+func newKey(keyType KeyType, privateKey internal.Raw, publicKey []byte, createdAt time.Time, metadata []byte) key {
+	return key{
+		keyType:    keyType,
+		privateKey: privateKey,
+		publicKey:  publicKey,
+		createdAt:  createdAt,
+		metadata:   metadata,
+	}
+}
+
+// EncryptionParams controls password-based encryption cost.
+// N and P are scrypt parameters; higher values increase CPU/memory cost.
+// Password is the secret used to derive the encryption key.
+type EncryptionParams struct {
+	Password     string
+	ScryptParams ScryptParams
+}
+
+func publicKeyFromPrivateKey(privateKeyBytes internal.Raw, keyType KeyType) ([]byte, error) {
+	switch keyType {
+	case Ed25519:
+		return ed25519.PublicKey(internal.Bytes(privateKeyBytes)), nil
+	case EcdsaSecp256k1:
+		// Here we use SEC1 (uncompressed) format for ECDSA public keys.
+		// Its commonly used and EVM addresses are derived from this format.
+		// We use the geth crypto library for secp256k1 support
+		// because stdlib doesn't support it.
+		privateKey, err := gethcrypto.ToECDSA(internal.Bytes(privateKeyBytes))
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert private key to ECDSA private key: %w", err)
+		}
+		pubKey := gethcrypto.FromECDSAPub(&privateKey.PublicKey)
+		return pubKey, nil
+	case X25519:
+		rv, err := curve25519.X25519(internal.Bytes(privateKeyBytes)[:], curve25519.Basepoint)
+		if err != nil {
+			return nil, fmt.Errorf("failed to derive shared secret: %w", err)
+		}
+		return rv, nil
+	default:
+		// Some types may not have a public key.
+		return []byte{}, nil
+	}
+}
+
+type keystore struct {
+	mu       sync.RWMutex
+	keystore map[string]key
+	storage  storage.Storage
+	enc      EncryptionParams
+}
+
+func NewKeystore(ctx context.Context, storage storage.Storage, enc EncryptionParams) (Keystore, error) {
```

Comment:
I know that context via constructor is frowned upon, but here I'm not sure - if I add a separate Load(ctx) function I think it'd add a bunch of boilerplate to clients always loading after creating and also error handling for every method checking if loaded or not. Seems simpler this way - wdyt @jmank88

Reply:
Usually the problems are deeper than this, like storing the context or launching a goroutine. This just seems like a matter of readability. We can keep concerns separate without forcing boilerplate on anyone though. In this case, the constructor is not doing anything else, so it could be eliminated in exchange for a load method:
```go
type StoredKeystore struct {
	Storage  storage.Storage
	Enc      EncryptionParams

	mu sync.RWMutex
	keystore map[string]key
}

func (k *StoredKeystore) Load(ctx context.Context) error {
```
Or something like an `Init` to avoid the exported fields:
```suggestion
func (*StoredKeystore) Init(ctx context.Context, storage storage.Storage, enc EncryptionParams) (Keystore, error) {
```

Reply:
Renamed. The concern was just about adding something like if not loaded: return ErrNotLoaded to all the ks methods to avoid potentially opaque errors on empty map